leakcanary-hprof/src/main/java/leakcanary/hprof/HProfParser.java

```diff
@@ -0,0 +1,1109 @@
+        return Arrays.asList(a);
+      }
+      case SHORT: {
+        Short[] a = new Short[numElements];
+        for (int i = 0; i < numElements; i++) {
+          a[i] = hprofSource.readShort();
+        }
+        return Arrays.asList(a);
+      }
+      case INT: {
+        Integer[] a = new Integer[numElements];
+        for (int i = 0; i < numElements; i++) {
+          a[i] = hprofSource.readInt();
+        }
+        return Arrays.asList(a);
+      }
+      case LONG: {
+        Long[] a = new Long[numElements];
+        for (int i = 0; i < numElements; i++) {
+          a[i] = hprofSource.readLong();
+        }
+        return Arrays.asList(a);
+      }
+      default:
+        throw new IllegalStateException("this can't happen");
+    }
+  }
+
+  public enum Type {
+    OBJECT(null, -1), // variable
+    BOOLEAN("boolean", 1),
+    CHAR("char", 2),
+    FLOAT("float", 4),
+    DOUBLE("double", 8),
+    BYTE("byte", 1),
+    SHORT("short", 2),
+    INT("int", 4),
+    LONG("long", 8);
+
+    public final String name;
+    public final int numBytes;
+
+    Type(String name, int numBytes) {
+      this.name = name;
+      this.numBytes = numBytes;
+    }
+
+    public static Type getType(byte type) {
+      switch (type) {
+        case 2:
+          return OBJECT;
+        case 4:
+          return BOOLEAN;
+        case 5:
+          return CHAR;
+        case 6:
+          return FLOAT;
+        case 7:
+          return DOUBLE;
+        case 8:
+          return BYTE;
+        case 9:
+          return SHORT;
+        case 10:
+          return INT;
+        case 11:
+          return LONG;
+        default:
+          throw new IllegalArgumentException("Unexpected type in heap dump: " + type);
+      }
+    }
+  }
+
+  private String readLineNumber(BufferedSource hprofSource) throws IOException {
+    int lineNumber = hprofSource.readInt();
+    switch (lineNumber) {
+      case 0:
+        return "no line information available";
+      case -1:
+        return "unknown location";
+      case -2:
+        return "compiled method";
+      case -3:
+        return "native method";
+      default:
+        return Integer.toString(lineNumber);
+    }
+  }
+
+  private int readSerialNumber(BufferedSource hprofSource) throws IOException {
+    int serialNumber = hprofSource.readInt();
+    //if (serialNumber <= 0) throw new IllegalStateException("serial number always > 0");
+    return serialNumber;
+  }
+
+  private long readId(BufferedSource hprofSource) throws IOException {
+    switch (idSize) {
+      case 4:
+        long id = hprofSource.readInt();
+        return id & 0x00000000ffffffffL;
+      case 8:
+        return hprofSource.readLong();
+      default:
+        throw new IllegalArgumentException("Invalid identifier size: " + idSize);
+    }
+  }
+
+  private String readUntilNull(BufferedSource hprofSource) throws IOException {
+    Buffer buffer = new Buffer();
+    byte b;
+    while ((b = hprofSource.readByte()) != 0) {
```

Comment:
Use `indexOf` so this loop can happen inside Okio

Reply:
nice!